contrib/drivers/huawei/dorado/metrics.go

```diff
@@ -0,0 +1,311 @@
+}
+
+// 	getMetricList:- is  to get the list of supported metrics for given resource type
+//	supportedMetrics -> list of supported metrics
+func (d *MetricDriver) GetMetricList(resourceType string) (supportedMetrics []string, err error) {
+	var configs Configs
+
+	//Read supported metric list from yaml config
+	source := []byte(data)
+	err = yaml.Unmarshal(source, &configs)
+	if err != nil {
+		log.Fatalf("unmarshal error: %v", err)
+	}
+
+	for _, resources := range configs.Cfgs {
+		if resources.Resource == resourceType {
+			for _, m := range resources.Metrics {
+				supportedMetrics = append(supportedMetrics, m)
+			}
+			break
+		}
+	}
+
+	return supportedMetrics, nil
+}
+
+func (d *MetricDriver) CollectPerformanceMetrics(resId string, metricList []string) (map[string]float64, error) {
+	name2id := map[string]string{
+		KMetricIOPS:               PerfIOPS,
+		KMetricBandwidth:          PerfBandwidth,
+		KMetricLatency:            PerfLatency,
+		KMetricServiceTime:        PerfServiceTime,
+		KMetricUtilizationPercent: PerfUtilizationPercent,
+		KMetricCacheHitRatio:      PerfCacheHitRatio,
+		KMetricCpuUsage:           PerfCpuUsage,
+	}
+
+	var idList = make([]string, len(metricList))
+	for i, name := range metricList {
+		idList[i] = name2id[name]
+	}
+
+	perfMap, err := d.client.GetPerformance(resId, idList)
+	if err != nil {
+		log.Errorf("get performance data failed: %s", err)
+		return nil, err
+	}
+
+	var metricMap = make(map[string]float64)
+	for _, name := range metricList {
+		v, _ := strconv.ParseFloat(perfMap[name2id[name]], 64)
+		metricMap[name] = v
+	}
+	return metricMap, nil
+}
+
+func (d *MetricDriver) CollectControllerMetrics() ([]*model.MetricSpec, error) {
+	// get Metrics to unit map
+	metricToUnitMap := getMetricToUnitMap(MetricResourceTypeController)
+	//validate metric support list
+	supportedMetrics, err := d.GetMetricList(MetricResourceTypeController)
+	if supportedMetrics == nil {
+		log.Infof("no metrics found in the supported metric list")
+	}
+	controllers, err := d.client.ListControllers()
+	if err != nil {
+		log.Errorf("get controller failed: %s", err)
+		return nil, err
+	}
+
+	var tempMetricArray []*model.MetricSpec
+	for _, controller := range controllers {
+		// TODO: the controller id need to be optional
+		uuid := ObjectTypeController + ":" + controller.Id
+		metricMap, err := d.CollectPerformanceMetrics(uuid, supportedMetrics)
+		if err != nil {
+			log.Errorf("get performance data failed: %s", err)
+			return nil, err
+		}
+
+		for _, element := range supportedMetrics {
+			metricValue := &model.Metric{
+				Timestamp: getCurrentUnixTimestamp(),
+				Value:     metricMap[element],
+			}
+			metricValues := make([]*model.Metric, 0)
+			metricValues = append(metricValues, metricValue)
+			metric := &model.MetricSpec{
+				InstanceID:   uuid,
+				InstanceName: uuid,
+				Job:          "HuaweiOceanStor",
+				Labels:       map[string]string{"device": MetricResourceTypeController},
```

Comment:
I suggest controller ID as the Label
Labels:       map[string]string{"controller": uuid}

Reply:
@NajmudheenCT done, BTW, can you tell me what is Lables used for?

Reply:
labels are to differentiate the characteristics of the thing that is being measured, That could be associated components/properties like controller,pool,port,storageGroup, etc..